src/libraries/System.Private.CoreLib/src/System/Collections/Generic/CollectionExtensions.cs

```diff
@@ -0,0 +1,164 @@
+        public static bool Remove<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            if (dictionary is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
+            }
+
+            if (dictionary.TryGetValue(key, out value))
+            {
+                dictionary.Remove(key);
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Returns a read-only <see cref="ReadOnlyCollection{T}"/> wrapper
+        /// for the specified list.
+        /// </summary>
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="list">The list to wrap.</param>
+        /// <returns>An object that acts as a read-only wrapper around the current <see cref="IList{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="list"/> is null.</exception>
+        public static ReadOnlyCollection<T> AsReadOnly<T>(this IList<T> list) =>
+            new ReadOnlyCollection<T>(list);
+
+        /// <summary>
+        /// Returns a read-only <see cref="ReadOnlyDictionary{TKey, TValue}"/> wrapper
+        /// for the current dictionary.
+        /// </summary>
+        /// <typeparam name="TKey">The type of keys in the dictionary.</typeparam>
+        /// <typeparam name="TValue">The type of values in the dictionary.</typeparam>
+        /// <param name="dictionary">The dictionary to wrap.</param>
+        /// <returns>An object that acts as a read-only wrapper around the current <see cref="IDictionary{TKey, TValue}"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="dictionary"/> is null.</exception>
+        public static ReadOnlyDictionary<TKey, TValue> AsReadOnly<TKey, TValue>(this IDictionary<TKey, TValue> dictionary) where TKey : notnull =>
+            new ReadOnlyDictionary<TKey, TValue>(dictionary);
+
+        /// <summary>Adds the elements of the specified span to the end of the <see cref="List{T}"/>.</summary>
+        /// <typeparam name="T">The type of elements in the list.</typeparam>
+        /// <param name="list">The list to which the elements should be added.</param>
+        /// <param name="source">The span whose elements should be added to the end of the <see cref="List{T}"/>.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="list"/> is null.</exception>
+        public static void AddRange<T>(this List<T> list, ReadOnlySpan<T> source)
+        {
+            if (list is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if (!source.IsEmpty)
+            {
+                if (list._items.Length - list._size < source.Length)
+                {
+                    list.Grow(list._size + source.Length);
```

Comment:
Per https://github.com/dotnet/runtime/issues/76116 this can potentially overflow (in this case it could happen if both `list._size` and `source.Length` equal `Array.MaxLength`)

```suggestion
                    list.Grow(checked(list._size + source.Length));
```

We could address the issue here, or I can pick it up after this PR gets merged.

Reply:
Adding checked is fine... but can you give an example where the values would pass the guard clause above and then overflow here?  Is this for a case where _size has been corrupted and might be > _items.Length?

Reply:
Oh, nevermind, I misread.  Yeah, checked is fine.

Reply:
This should probably added to `InsertRange` as well.